### PR TITLE
transparent_user

### PR DESCRIPTION
**Pull request by ilejn**
The PR is superceeded by https://github.com/ContentSquare/chproxy/pull/219 , should be closed
 
Implements https://github.com/ContentSquare/chproxy/issues/169

Flowchart https://hackmd.io/@sPX7lB-2R7SNbZ7C7xpbKQ/HkqVB8J5c

Wildcarded users concept https://hackmd.io/LXk8tLI8Q3etKax-HWtGxQ

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -514,6 +518,9 @@ type User struct {
 	// Name of Cache configuration to use for responses of this user
 	Cache string `yaml:"cache,omitempty"`
 
+	// Whether to allow cache for transparent_user
```
Comment (aosorgin) [inline config/config.go:521]: Something happened to the indents

File: config/examples/transparent.yml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+server:
+  http:
+      listen_addr: ":9090"
+      allowed_networks: ["127.0.0.0/24"]
+
+users:
+  - name: "transparent_user"
+    to_cluster: "default_with_transparent"
+    to_user: "transparent_user"
```
Comment (aosorgin) [inline config/examples/transparent.yml:11]: allow_transparent_cache: true ?

File: proxy.go [hunk 3]
```diff
@@ -518,12 +518,30 @@ func (rp *reverseProxy) getScope(req *http.Request) (*scope, int, error) {
...
 		cu = c.users[u.toUser]
 	}
+
+	isTransparent := false
+	if u == nil {
+		u = rp.users[config.TransparentUser]
+		if u != nil {
+			c = rp.clusters[u.toCluster]
+			if c != nil {
+				transparent_cu := *c.users[u.toUser]
```
Comment (aosorgin) [inline proxy.go]: Let's check if c.users[u.toUser] is not nil

File: scope.go [hunk 4]
```diff
@@ -744,7 +744,11 @@ func newCluster(c config.Cluster) (*cluster, error) {
 		clusterUsers[cu.Name] = newClusterUser(cu)
 	}
 
-	heartBeat := newHeartBeat(c.HeartBeat, c.ClusterUsers[0])
+	heartBeatClusterUser := c.ClusterUsers[0]
```
Comment (aosorgin) [inline scope.go:747]: This code is duplicated. Maybe it can be moved to newHeartBeat? Moreover, should we check if transparent user is allowed?

File: config/examples/transparent.yml (hunk 2 above)
Comment (ilejn) [reply config/examples/transparent.yml:11]: There is no cache in this config, so the value does not matter, while it is false by default.
The use case where a transparent user exists and the cache enabled is covered by bad.transparent_user_with_cache.yml and correspondent test

File: proxy.go (hunk 3 above)
Comment (ilejn) [reply proxy.go]: This is already checked, see comment above that says "c and cu for toCluster and toUser must exist if applyConfig is correct. Fix applyConfig if c or cu equal to nil."

Actually I would have added an assert if I knew that a concept of assert exists in golang ;)  Looking into this.

File: scope.go (hunk 4 above)
Comment (ilejn) [reply scope.go:747]: Yes, it is replicated in a test. Actually it is what tests are needed for.
Moving into newHeartBeat  ... well, possible but not sure.

File: config/config.go (hunk 1 above)
Comment (ilejn) [reply config/config.go:521]: actually everything is Ok, no format issues

File: config/examples/transparent.yml [hunk 5]
```diff
@@ -0,0 +1,19 @@
+log_debug: true
```
Comment (gontarzpawel) [inline config/examples/transparent.yml:1]: Isn't this file missing `allow_transparent_cache: true` ?

File: docs/content/en/use-cases/transparent.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
+---
+title: Pass user names and passwords as is
+menuTitle: Transpaent users
```
Comment (gontarzpawel) [inline docs/content/en/use-cases/transparent.md]:
```suggestion
menuTitle: Transparent users
```

File: docs/content/en/use-cases/transparent.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    to_cluster: "default_with_transparent"
+    to_user: "transparent_user"
+
+# by default each cluster has `default` user which can be overridden by section `users`
+clusters:
+  - name: "default_with_transparent"
+    nodes: ["127.0.0.1:8123"]
+
+    users:
+    - name: "transparent_user"
```
Comment (gontarzpawel) [inline docs/content/en/use-cases/transparent.md:29]: As well here, according to the logic this is not the correct configuration . Transparent user should be enabled

File: heartbeat_test.go [hunk 8]
```diff
@@ -70,7 +70,11 @@ func TestNewHeartBeat(t *testing.T) {
 	}
 	testCompareNum(t, "cluster.heartbeat.interval", int64(c.heartBeat.interval/time.Microsecond), int64(time.Duration(5*time.Second)/time.Microsecond))
 
-	hb := newHeartBeat(heartBeatFullCfg, clusterCfg.ClusterUsers[0])
+	heartBeatClusterUser := clusterCfg.ClusterUsers[0]
+	if heartBeatClusterUser.Name == config.TransparentUser {
+		heartBeatClusterUser = config.DefaultClusterUser
+	}
+	hb := newHeartBeat(heartBeatFullCfg, heartBeatClusterUser)
```
Comment (gontarzpawel) [inline heartbeat_test.go:77]: Why is this change needed?

File: proxy.go (hunk 3 above)
Comment (gontarzpawel) [inline proxy.go]: let's stick to the same naming convention everywhere
```suggestion
				transparentCu := *c.users[u.toUser]
```

File: proxy.go [hunk 9]
```diff
@@ -518,12 +518,30 @@ func (rp *reverseProxy) getScope(req *http.Request) (*scope, int, error) {
 		c = rp.clusters[u.toCluster]
 		cu = c.users[u.toUser]
 	}
+
+	isTransparent := false
+	if u == nil {
```
Comment (gontarzpawel) [inline proxy.go]: This nil check is already applied in line 541, why don't we move this block of code after that line (together with `rp.lock.RUnlock()` )? 

Would you mind also extract this logic to the dedicated function? Otherwise `getScope` becomes too complex to read

File: scope.go (hunk 4 above)
Comment (gontarzpawel) [inline scope.go:747]: Taking occasion of your change, I don't understand what's the motivation of using the first user defined in the list for cluster. I think it can create a great confusion, which we actually already experienced in Contentsquare. Let's assign by default a default user for heartbeats. Could you do that?

**Review by gontarzpawel — CHANGES_REQUESTED** [review]
Thanks for your contribution!

File: config/examples/transparent.yml (hunk 5 above)
Comment (ilejn) [reply config/examples/transparent.yml:1]: Until a configuration does not use cache, allow_transparent_cache is not needed.
I assume that while cache is not configured, it is disabled. Is it wrong?

File: docs/content/en/use-cases/transparent.md (hunk 7 above)
Comment (ilejn) [reply docs/content/en/use-cases/transparent.md:29]: There is no notion of enabling transparent user in current implementation.
If a user with magic name 'transparent_user' is listed, It is enabled.
If you believe that this behaviour is confusing and we need an explicit setting to enable transparent users - lets discuss.
'allow_transparent_cache' is for cache only.

File: heartbeat_test.go (hunk 8 above)
Comment (ilejn) [reply heartbeat_test.go:77]: Neither 'transparent_user' nor a user from request credentatils are known for ClickHouse, that is why we have to use another user, e.g. default.

File: scope.go (hunk 4 above)
Comment (ilejn) [reply scope.go:747]: It is common to disable a user _default_ in ClickHouse completely, so we cannot use _default_ for heartbeat unconditionally.
Could you elaborate on ' assign by default a default user for heartbeats'?
Do you suggest  we should introduce clusters/heartbeat/name (and probably clusters/heartbeat/password), and if it is not specified use 'deafult'?

File: proxy.go (hunk 9 above)
Comment (ilejn) [reply proxy.go]: The checks in 523 and in 541 are different. 
The logic here is
```
if user_is_not_set:
    attempt_to_use_transparent_user
if user_is_not_set:   # even transparent logic did not work
    return error
```
To be honest I do not see obvious issues here, while I
* add a comment or a couple of ones
* Ok, I'll try to move this logic to a function

File: config/examples/transparent.yml (hunk 5 above)
Comment (gontarzpawel) [reply config/examples/transparent.yml:1]: oh I misunderstood. What's the advantage of introducing `allow_transparent_cache` then? If user wants to add caching he will need to link it in the config section as usual, if not otherwise.

Comment (ilejn) [reply config/examples/transparent.yml:1]: The idea of transparent_user is to represent several real CH users.
These users would share cache with each other.
I consider this as a potential security risk e.g. because of Row Level Security malfunctioning.
That is why I've decided to make this risk more obvious, _allow_transparent_cache_ is somewhat similar to _hack_me_please._

File: scope.go (hunk 4 above)
Comment (gontarzpawel) [reply scope.go:747]: I didn't know default user is being disabled ^^ it's a valid point. I will create a separate issue to add a specific user for heartbeats and fallback to the default one if not provided

**Comment by ilejn** [comment]
@Garnek20 , there are some indentation changes in config_test.go (silently introduced by my emacs).
Should I keep them as they are in my branch or it is better to rollback?

File: proxy.go (hunk 9 above)
Comment (ilejn) [reply proxy.go]: @Garnek20  , I've introduced getTransparentUser and getUser.
Not sure that it makes sense, it seems to look too heavy.

Although it is not that bad as getTransparentUser together with original logic in getScope.

File: config/examples/transparent.yml (hunk 5 above)
Comment (gontarzpawel) [reply config/examples/transparent.yml:1]: What do you think about adding a config parameter to the user section explicitly specifying if it's an ldap clickhouse user? This way no such mechanism would be needed (I think). It's a bit unintuitive to me

Comment (ilejn) [reply config/examples/transparent.yml:1]: Not definitely LDAP, maybe just Kerberos. Another possible concern - plain passwords in chproxy config someone would like to avoid.
So, any user we would like to auth using ClickHouse, not chproxy facilities.

Using a magic name does not seem perfect, agree, but I do not fully understand your idea.

If you suggest something like
```
users:
  - name: "some_user"
    to_cluster: "default_with_transparent"
    to_user: "some_user"
    transparency: true
clusters:
  - name: "default_with_transparent"
    nodes: ["127.0.0.1:8123"]

    users:
    - name: "some_user"
      transparency: true
```
it is doable of course, we do not use a magic name anymore, but use an explicit flag. 
Is it better from your point of view?

The whole idea is to be able to pass original credentials to ClickHouse for some (not all) users.
It is impossible to have the real usernames of such users in chproxy config.
That is why we need two types of users in chproxy configuration anyway.

**Comment by ilejn** [comment]
Hello @Garnek20 , have a look at the diagram  https://hackmd.io/@sPX7lB-2R7SNbZ7C7xpbKQ/HkqVB8J5c
Hope that it clarifies the whole idea.

File: config/examples/transparent.yml (hunk 5 above)
Comment (mga-chka) [reply config/examples/transparent.yml:1]: did you think of creating an LDAP connector in order to fetch the users instead of creating this transparent_user concept that breaks a bit the concept of users in CHProxy (cf the issues with caching)?

File: proxy.go [hunk 10]
```diff
@@ -499,6 +499,42 @@ func (rp *reverseProxy) refreshCacheMetrics() {
...
+	defer rp.lock.RUnlock()
+
+	*u = rp.users[config.TransparentUser]
+	if *u != nil {
+		*c = rp.clusters[(*u).toCluster]
+		if *c != nil {
+			transparentCu := (*c).users[(*u).toUser]
+			if transparentCu != nil {
+				*cu = transparentCu
+				(*cu).name = name
```
Comment (mga-chka) [inline proxy.go:529]: There is one thing that worries me with this transparent user concept.

What if the user has a default CH user? I know that it's a bad habit to keep it but I bet many companies are keeping it and some of them without passwords. If the default user has a pwd then CH proxy is a door to bruteforce the pwd. If the default user doesn't have pwd then it's open bar.

So you should return false if the cluster user name is empty

File: config/examples/transparent.yml (hunk 5 above)
Comment (ilejn) [reply config/examples/transparent.yml:1]: Hello Christophe, 
I am not sure that I fully understand the idea: if a user is not listed in chproxy config, how his or her properties (e.g. quotas) can be defined? It does not look much better than transparent_user, but definitely more complex.

I agree with you, transparent user concept is not in line with the concept of chproxy users. 
But we should face reality, native ClickHouse authentication facilities are rather advanced.
They include
- LDAP support. There are several options, e.g Role Based Access Control.
- Kerberos support (which is a special problem).
- native user management via SQL with passwords in sha256

So, it is not just about LDAP. 
Speaking of which, setting up LDAP is rather cumbersome sometimes, and the idea of doing it twice is not necessarily attractive for Operations.

What can be done to make chproxy users great again? 
For example, we can introduce wildcards. 
Instead of a single magic transparent_user we can have users like *_dba , *_departmet1_analyst, *_department2_analyst. These transparent (or umbrella or whatever called) users can be maintained by chproxy from quotas and cache point of view, leaving real authenticating and authorization for ClickHouse. 
This approach seems to be safe and convenient (unless I am missing something).
@mga-chka , what do you think?

File: proxy.go (hunk 10 above)
Comment (ilejn) [reply proxy.go:529]: Agree.
BTW, the risk of brute force is mitigated by the idea of wildcards that is described in my comment above https://github.com/ContentSquare/chproxy/pull/170#discussion_r903084145 .

File: config/examples/transparent.yml (hunk 5 above)
Comment (ilejn) [reply config/examples/transparent.yml:1]: Prefixes might be better than suffixes, dba_*, not *_dba

File: docs/content/en/use-cases/transparent.md (hunk 6 above)
Comment (ilejn) [reply docs/content/en/use-cases/transparent.md]: Fixed

File: proxy.go (hunk 3 above)
Comment (ilejn) [reply proxy.go]: Fixed

File: config/examples/transparent.yml (hunk 5 above)
Comment (mga-chka) [reply config/examples/transparent.yml:1]: Sorry for not answering now. It just to let you know that we're discussing with pawel about your PR (because it's conceptually impactful) but since we're busy on other tasks we might give an answer at the beginning of next week.

Comment (ilejn) [reply config/examples/transparent.yml:1]: Hello Christophe, yes, I understand that it is rather serious change and it should be properly discussed and actually properly designed.
I believe that the problem is important for potential and current chproxy users and it should be solved in one or another way. Quick-fix approach like in this PR may be not the best, while whildcards look Ok to me.

We (Arenadata) are thinking about switching to chproxy from our current inhouse product, that is why we need some extra functionality. And we are ready to invest our time in it of course.

File: heartbeat_test.go (hunk 8 above)
Comment (ilejn) [reply heartbeat_test.go:77]: Sorry, the statement "... a user from request credentatils are known for ClickHouse" does not make sense, no client's request in heartbeat context.

File: config/examples/transparent.yml (hunk 5 above)
Comment (mga-chka) [reply config/examples/transparent.yml:1]: Hi, 

We believe that having a specific LDAP support inside CHProxy is the cleaner way regarding the current logic of CH proxy (even if it means duplicating the logic of CH and doing the same for kerberos and so on).
That being said, you're right that CHProxy is far from what CH provides in terms of user management.

So, here is our proposal: having a specific config flag that will deactivate ALL the user logic and the features depending on users (caching, quotas). In this mode, CHProxy would be just a simple proxy whose job is to reach on off the clusters nodes.
Letting people use users based features without the user concepts will for sure lead to bugs so this is why we don't want that for the moment.

Nb: if we rework the user management, nothing prevent us in the future to activate cleanly the user based features even if the users are managed only by CH

WDYT?

Comment (ilejn) [reply config/examples/transparent.yml:1]: Hello Christophe,
I've added a description of wildcarded users concept to the PR https://hackmd.io/LXk8tLI8Q3etKax-HWtGxQ .
To me, it is not that bad, actually, it has some advantages.
On the other hand, your concept looks solid, it is unbeatable. Maybe it is just too perfect for our non-perfect world? ;)
What about having the flag that disables user machinery completely AND having wildcarded users to let chproxy customers decide if they take risks of 'impersonation' or not.?

Comment (mga-chka) [reply config/examples/transparent.yml:1]: Hi, I left the following comments a few days ago in your documents 
"[christophe] what about the caching part? Do we assume that like for quota it will be shared by all the department?"
I'm assuming the answer will be yes. If so then ok, let's do your proposal.

Comment (mga-chka) [reply config/examples/transparent.yml:1]: since you need to rewrite a part of your PR, can you please avoid the use of "**" (for example in getUser() or getTransparentUser() functions)?
I know we're picky but we want to keep the code consistent and no one has used a pointer of pointer in chproxy codebase until now.
nb: We will soon add code guidelines so that it will be more explicit

Comment (ilejn) [reply config/examples/transparent.yml:1]: Hello Christophe,
thank you for your response.
Yes, I assume the same, the cache "will be shared by all the department".
Cache works for a wildcarded_user/umbrella _user/department/whatever_it_called in the same way as for normal user.

So we are going with wilcarded user concept. 
Cool!
Should we create another PR or it is Ok to continue in current PR and current branch?

Comment (mga-chka) [reply config/examples/transparent.yml:1]: If you can, create a new PR because the code will change a lot and this PR already has a lot of comments. If it takes you too much time keep this PR, it just that it will more difficult to read if someone wants to understand the creation of the transparent user feature

Comment (ilejn) [reply config/examples/transparent.yml:1]: Hello Christophe @mga-chka , sorry for the delay. 
Vacation is over and I am back to work.
Could you help me with the name of these new users, please? Is 'wildcarded user' Ok?

Comment (ilejn) [reply config/examples/transparent.yml:1]: Hello @mga-chka , could you help me with the name of these new users, please?
Going with "wildcarded user" so far.

Comment (mga-chka) [reply config/examples/transparent.yml:1]: my bad, I did see this notification then I was off. Yes let's go for this name (since I don't have something better)

**Comment by ilejn** [comment]
New PR https://github.com/ContentSquare/chproxy/pull/219 created according to @mga-chka suggestion.
